File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/L._Полный_граф/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>


void solution() {
    /*
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    */
    int n, m;
    std::cin >> n >> m;
    std::unordered_map<int, std::unordered_set<int>> adjList;
    for (int i = 0; i < m; ++i) {
        int u, v;
        std::cin >> u >> v;
        if (u != v) {
            adjList[u].insert(v);
            adjList[v].insert(u);
        }
    }
    for (auto& p : adjList) {
        if (p.second.size() != n - 1) {
            std::cout << "NO\n";
            return;
        }
    }
    std::cout << ((adjList.size() == n) || (n == 1)? "YES" : "NO") << "\n";
}


int main() {
    solution();
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/L._Полный_граф/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    adj_list: dict[int, set[int]] = dict()
    for _ in range(m):
        u, v = map(int, sys.stdin.readline().rstrip().split())
        if u != v:
            if u not in adj_list: adj_list[u] = set()
            adj_list[u].add(v)
            if v not in adj_list: adj_list[v] = set()
            adj_list[v].add(u)
    for vertex, edges in adj_list.items():
        if len(edges) != n - 1:
            sys.stdout.write('NO\n')
            return
    sys.stdout.write('YES\n' if (len(adj_list) == n or n == 1) else 'NO\n')


if __name__ == '__main__':
    solution()
